[ppc64] ICE: Region parameter out of range when substituting in region'b

Trying to natively build powerpc64 as of 380100c568c1c390c837caf98338d07c4d8a0462, I get an ICE in stage 1:

```text
Copying stage1 compiler (powerpc64-unknown-linux-gnu)
Building stage1 std artifacts (powerpc64-unknown-linux-gnu -> powerpc64-unknown-linux-gnu)
   Compiling core v0.0.0 (file:///home/jistone/rust/src/libcore)
error: internal compiler error: src/librustc/ty/subst.rs:411: Region parameter out of range when substituting in region'b (root type=Some((&&'b str,))) (index=1)

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.20.0-dev (380100c56 2017-06-20) running on powerpc64-unknown-linux-gnu

note: run with `RUST_BACKTRACE=1` for a backtrace

thread 'rustc' panicked at 'Box<Any>', src/librustc_errors/lib.rs:426
stack backtrace:
   0: rust_metadata_std_1ef34ef8b50021bf3e2b2bca3dfc146e
   1: rust_metadata_std_1ef34ef8b50021bf3e2b2bca3dfc146e
   2: rust_metadata_std_1ef34ef8b50021bf3e2b2bca3dfc146e
   3: rust_metadata_std_1ef34ef8b50021bf3e2b2bca3dfc146e
   4: rust_metadata_std_1ef34ef8b50021bf3e2b2bca3dfc146e
   5: rust_metadata_rustc_e1561b4cf34be67f9773866f88239afd
   6: rust_metadata_rustc_e1561b4cf34be67f9773866f88239afd
   7: rust_metadata_rustc_e1561b4cf34be67f9773866f88239afd
   8: rust_metadata_rustc_e1561b4cf34be67f9773866f88239afd
   9: rust_metadata_rustc_e1561b4cf34be67f9773866f88239afd
  10: rust_metadata_rustc_e1561b4cf34be67f9773866f88239afd
  11: rust_metadata_rustc_e1561b4cf34be67f9773866f88239afd
  12: rust_metadata_rustc_e1561b4cf34be67f9773866f88239afd
  13: rust_metadata_rustc_e1561b4cf34be67f9773866f88239afd
  14: rust_metadata_rustc_e1561b4cf34be67f9773866f88239afd
  15: rust_metadata_rustc_e1561b4cf34be67f9773866f88239afd
  16: rust_metadata_rustc_e1561b4cf34be67f9773866f88239afd
  17: rust_metadata_rustc_e1561b4cf34be67f9773866f88239afd
  18: rust_metadata_rustc_e1561b4cf34be67f9773866f88239afd
  19: rust_metadata_rustc_e1561b4cf34be67f9773866f88239afd
  20: rust_metadata_rustc_e1561b4cf34be67f9773866f88239afd
  21: <unknown>
  22: <unknown>
  23: rust_metadata_rustc_e1561b4cf34be67f9773866f88239afd
  24: rust_metadata_rustc_e1561b4cf34be67f9773866f88239afd
  25: rust_metadata_rustc_e1561b4cf34be67f9773866f88239afd
  26: rust_metadata_rustc_e1561b4cf34be67f9773866f88239afd
  27: rust_metadata_rustc_e1561b4cf34be67f9773866f88239afd
  28: rust_metadata_rustc_e1561b4cf34be67f9773866f88239afd
  29: rust_metadata_rustc_e1561b4cf34be67f9773866f88239afd
  30: rust_metadata_rustc_e1561b4cf34be67f9773866f88239afd
  31: rust_metadata_rustc_e1561b4cf34be67f9773866f88239afd
  32: rust_metadata_rustc_e1561b4cf34be67f9773866f88239afd
  33: rust_metadata_rustc_e1561b4cf34be67f9773866f88239afd
  34: rust_metadata_rustc_e1561b4cf34be67f9773866f88239afd
  35: rust_metadata_rustc_e1561b4cf34be67f9773866f88239afd
  36: rust_metadata_rustc_e1561b4cf34be67f9773866f88239afd
  37: <unknown>
  38: <unknown>
  39: <unknown>
  40: <unknown>
  41: <unknown>
  42: <unknown>
  43: rust_metadata_std_1ef34ef8b50021bf3e2b2bca3dfc146e
  44: <unknown>
  45: rust_metadata_std_1ef34ef8b50021bf3e2b2bca3dfc146e
  46: <unknown>
  47: <unknown>

error: Could not compile `core`.
```

I got an s390x build, and it fared even worse.  Tons of move errors on `Copy` types like this:

```
error[E0382]: use of moved value: `self`
    --> src/libcore/num/mod.rs:1412:28
     |
1412 |               (self << n) | (self >> (($BITS - n) % $BITS))
     |                ----          ^^^^ value used here after move
     |                |
     |                value moved here
...
2261 | /     uint_impl! { u16, u16, 16,
2262 | |         intrinsics::ctpop,
2263 | |         intrinsics::ctlz,
2264 | |         ctlz_nonzero,
...    |
2268 | |         intrinsics::sub_with_overflow,
2269 | |         intrinsics::mul_with_overflow }
     | |_______________________________________- in this macro invocation
     |
     = note: move occurs because `self` has type `u16`, which does not implement the `Copy` trait
```

before it finally ICEd:

    thread 'rustc' panicked at 'assertion failed: self.qualif.intersects(Qualif::MUTABLE_INTERIOR)', src/librustc_mir/transform/qualify_consts.rs:732

I'm guessing that these particular errors and ICEs are not significant in themselves, only as signs of some more general corruption.

--- a/src/libcore/ptr.rs
+++ b/src/libcore/ptr.rs
@@ -374,10 +374,7 @@ unsafe fn swap_nonoverlapping_bytes(x: *mut u8, y: *mut u8, len: usize) {
     // #[repr(simd)], even if we don't actually use this struct directly.
     //
     // FIXME repr(simd) broken on emscripten and redox
-    // It's also broken on big-endian powerpc64 and s390x.  #42778
-    #[cfg_attr(not(any(target_os = "emscripten", target_os = "redox",
-                       target_endian = "big")),
-               repr(simd))]
+    #[cfg_attr(not(any(target_os = "emscripten", target_os = "redox")), repr(simd))]
     struct Block(u64, u64, u64, u64);
     struct UnalignedBlock(u64, u64, u64, u64);
 

